Sort app scripts by dependencies

mejo- opened the pull request:
There must be a smarter and less expensive way to implement this, it's
just a first attempt.

Fix #30278

| unsorted | sorted |
|:---------:|:------:|
|![2021-12-16_08-36](https://user-images.githubusercontent.com/14975046/146328088-a51d38d1-6b00-43d1-a577-18293782653e.png)|![2021-12-16_08-36_1](https://user-images.githubusercontent.com/14975046/146328103-43e7618e-cbff-4297-be6b-64b42635dd68.png)|

mejo- committed:
Sort app scripts by dependencies (Fixes: #30278)

Signed-off-by: Jonas Meurer <jonas@freesources.org>

skjnldsv commented on lib/public/Util.php:
```diff
@@ -241,10 +246,28 @@ public static function getScripts(): array {
+
+		// Sort app scripts by dependency
+		$sortedScripts = [];
+		foreach ($appScripts as $app => $scripts) {
+			if (array_key_exists($app, self::$scriptDeps)) {
+				$dep = self::$scriptDeps[$app];
+				if (!array_key_exists($dep, $sortedScripts)) {
+					$sortedScripts[$dep] = [];
+				}
+				$sortedScripts[$dep] += array_merge($sortedScripts[$dep], $scripts);
+			} else {
+				if (!array_key_exists($app, $sortedScripts)) {
+					$sortedScripts[$app] = [];
+				}
+				$sortedScripts[$app] = array_merge($scripts, $sortedScripts[$app]);
+			}
+		}
+
 		// sort core first
-		$scripts = array_merge(isset($appScripts['core']) ? $appScripts['core'] : [], ...array_values($appScripts));
+		$finalScripts = array_merge(isset($sortedScripts['core']) ? $sortedScripts['core'] : [], ...array_values($sortedScripts));
 		// remove duplicates
-		return array_unique($scripts);
+		return array_unique($finalScripts);
```
```suggestion
		$sortByDeps = function(string $app1, string $app2): int {
			// Always sort core first
			if ($app1 === 'core') {
				return -1;
			}
			if ($app2 === 'core') {
				return 1;
			}

			// If app1 depends on app2
			if (array_key_exists($app1, self::$scriptDeps) && self::$scriptDeps[$app1] === $app2) {
				return 1;
			}
			// If app2 depends on $app1
			if (array_key_exists($app2, self::$scriptDeps) && self::$scriptDeps[$app2] === $app1) {
				return -1;
			}
			return 0;
		};
		uksort($appScripts, $sortByDeps);

		// flatten array and remove duplicates 
		return array_unique(array_merge(...array_values($appScripts)));
```

skjnldsv replied:
See my suggestion for simpler and cleaner code!
Nice job storing the deps, the logic makes much more sense! :rocket:

skjnldsv commented:
Something I've been wondering, shall we make all the translations depends on core?
Like we init the base layer, then all translations, then apps?

Since apps can have their scripts loaded far before their translations, I feel this is safer:
```php
// addTranslations
self::$scripts['core']['last'][] = $path;
```

skjnldsv reviewed (changes requested):
```php
	/**
	 * Return the list of scripts injected to the page
	 * @return array
	 * @since 24.0.0
	 */
	public static function getScripts(): array {
		// Merging first and last data set
		$mapFunc = function (array $scriptsArray): array {
			return array_merge(...array_values($scriptsArray));
		};
		$appScripts = array_map($mapFunc, self::$scripts);

		// Sort by dependency if any
		$sortByDeps = function(string $app1, string $app2): int {
			// Always sort core first
			if ($app1 === 'core') {
				return -1;
			}
			if ($app2 === 'core') {
				return 1;
			}

			// If app1 depends on app2
			if (array_key_exists($app1, self::$scriptDeps) && self::$scriptDeps[$app1] === $app2) {
				return 1;
			}
			// If app2 depends on $app1
			if (array_key_exists($app2, self::$scriptDeps) && self::$scriptDeps[$app2] === $app1) {
				return -1;
			}
			return 0;
		};
		uksort($appScripts, $sortByDeps);

		// Flatten array and remove duplicates 
		return array_unique(array_merge(...array_values($appScripts)));
	}
```

Here is my suggestion for simpler and cleaner code!
Nice job storing the deps, the logic makes much more sense! :rocket:

skjnldsv commented:
Please not that might create some unsupported setups though.
Like scripts can depends on various different apps
But we store the deps per app, and not per script.

So if I have the following: 
```php
Util::addScript('viewer, 'viewer-main', 'files');
Util::addScript('text, 'text-main', 'viewer');
// for some reason
Util::addScript('text, 'text-public', 'core');
```

Then I might have `text` put after `core` and not after `viewer`

mejo- commented:
> Here is my suggestion for simpler and cleaner code!

Thanks @skjnldsv! Unfortunately your solution doesn't fix the problem. Consider a situation with `$appScripts` containing apps `[one, four, three, two]`, and app `four` depending on `two`. Since they're never directly compared, they won't be exchanged by `sortByDeps` :cry:

come-nc commented on lib/public/Util.php:
```diff
@@ -241,10 +246,28 @@ public static function getScripts(): array {
+				if (!array_key_exists($dep, $sortedScripts)) {
+					$sortedScripts[$dep] = [];
+				}
+				$sortedScripts[$dep] += array_merge($sortedScripts[$dep], $scripts);
```
Is `+=` on purpose? It is redundant with the `array_merge`, no?

mejo- replied:
You're right, good catch :blush:

skjnldsv commented:
> Since they're never directly compared, they won't be exchanged by `sortByDeps`

Nice catch. Good enough?

```php

	/**
	 * Return the list of scripts injected to the page
	 * @return array
	 * @since 24.0.0
	 */
	public static function getScripts(): array {
		// Merging first and last data set
		$mapFunc = function (array $scriptsArray): array {
			return array_merge(...array_values($scriptsArray));
		};
		$appScripts = array_map($mapFunc, self::$scripts);

		// Sort by dependency if any
		$sortedScripts = [];
		foreach(array_keys($appScripts) as $app) {
			// If we already processed the scripts by a dependency, continue
			if (array_key_exists($app, $sortedScripts)) {
				continue;
			}

			// If app depends on another one, append the dependency scripts first
			if (array_key_exists($app, self::$scriptDeps)) {
				$dep = self::$scriptDeps[$app];
				$sortedScripts[] = $appScripts[$dep];
			}
			// Append scripts
			$sortedScripts[] = $appScripts[$app];
		}

		// Flatten array and remove duplicates 
		return array_unique(array_merge(...array_values($sortedScripts)));
	}
```

come-nc commented:
`isset` is both faster and clearer than `array_key_exists`, no?

skjnldsv commented:
> `isset` is both faster and clearer than `array_key_exists`, no?

https://gist.github.com/alcaeus/536156663fac96744eba77b3e133e50a
```
$ php in_array_vs_isset_vs_array_key_exists.php 
Function call in_array took 1.51871 seconds
Function call isset took 0.14684 seconds
Function call array_key_exists took 0.22123 seconds
```

mejo- commented:
> > Since they're never directly compared, they won't be exchanged by `sortByDeps`
> 
> Nice catch. Good enough?

Sorry to say, but still doesn't solve the problem. Imagine the following graph:

`self::$scripts`:
```yaml
- text:
    first:
      - text/l10n/en
    last: []
- viewer:
    first:
      - viewer/l10n/en
    last:
      - text/js/text-viewer
- files:
    first: []
    last:
      - viewer/js/viewer-main
```

`self::$scriptDeps`:
```yaml
- text: viewer
- viewer: files
```

Your approach will result in:

```yaml
- viewer:
    - viewer/l10n/en
    - text/js/text-viewer
- text:
    - text/l10n/en
- files:
    - viewer/js/viewer-main
```

Files, which holds the viewer script (that is the real dependency for the text script) is still the last in order.

mejo- commented:
I pushed an updated version of my solution, using `array_push` instead of `array_merge` for better performance.

skjnldsv commented:
Thanks for tackling this, I'm not in the best algo mindset lately it seems :)

mejo- committed:
Improve unit testing for Util::addScript() and Util::getScript()

Instead of checking for a predefined order of the scripts, test the
logic: core first, dependencies before their children, no duplicates and
all scripts still listed.

Signed-off-by: Jonas Meurer <jonas@freesources.org>

mejo- committed:
Further improve addScript logic, migrate to uksort()

Instead of the logic with `first` and `last`, store dependencies in an
own array and sort the scripts topologically by its dependencies later.

Signed-off-by: Jonas Meurer <jonas@freesources.org>

mejo- commented:
Well, after further thinking this through, I found a flaw in my implementation - and as it turned out in the whole `first`/`last` logic.

I ended up adding proper unit tests to check for all the logic that we want: https://github.com/nextcloud/server/pull/30279/commits/3cef72e85354c843386e5f37b6b2ce5b73882d44

And then I came up with a solution that hopefully implements proper topological sorting for the scripts by its dependencies. I removed the former `first`/`last` logic in `addScript()` and instead just store the dependencies in an own array. In `getScripts()` I sort the scripts using `uksort()`. In the `uksort()` callable, I check whether the script is still before its dependency and in this case move it backwards:

```php
		$sortByDeps = static function (string $app1, string $app2): int {
			// Always sort core first
			if ($app1 === 'core') {
				return -1;
			}
			if ($app2 === 'core') {
				return 1;
			}

			// If app1 has a dependency
			if (array_key_exists($app1, self::$scriptDeps)) {
				$apps = array_keys(self::$scripts);
				// Move app1 backwards if dep comes afterwards
				if (array_search($app1, $apps, true) <
					array_search(self::$scriptDeps[$app1], $apps, true)) {
					return 1;
				}
			}

			return 0;
		};
		uksort(self::$scripts, $sortByDeps);

		// Flatten array and remove duplicates
		return array_unique(array_merge(...array_values(self::$scripts)));
```

juliushaertl commented on lib/public/Util.php:
```diff
@@ -236,15 +212,32 @@ public static function addScript($application, $file = null, $afterAppId = null)
+		uksort(self::$scripts, $sortByDeps);
+
+		// Flatten array and remove duplicates
+		return array_unique(array_merge(...array_values(self::$scripts)));
```
This line causes CI failure on installation/maintenance mode:

```
{
  "Exception": "Error",
  "Message": "array_merge() expects at least 1 parameter, 0 given at /drone/src/lib/public/Util.php#240",
  "Code": 0,
  "Trace": [
    {
      "function": "onError",
      "class": "OC\\Log\\ErrorHandler",
      "type": "::",
      "args": [
        2,
        "array_merge() expects at least 1 parameter, 0 given",
        "/drone/src/lib/public/Util.php",
        240,
        {
          "sortByDeps": {
            "__class__": "Closure"
          }
        }
      ]
    },
    {
      "file": "/drone/src/lib/public/Util.php",
      "line": 240,
      "function": "array_merge",
      "args": []
    },
    {
      "file": "/drone/src/lib/private/TemplateLayout.php",
      "line": 212,
      "function": "getScripts",
      "class": "OCP\\Util",
      "type": "::",
      "args": []
    },
    {
      "file": "/drone/src/lib/private/legacy/OC_Template.php",
      "line": 177,
      "function": "__construct",
      "class": "OC\\TemplateLayout",
      "type": "->",
      "args": [
        "guest",
        ""
      ]
    },
    {
      "file": "/drone/src/lib/private/Template/Base.php",
      "line": 132,
      "function": "fetchPage",
      "class": "OC_Template",
      "type": "->",
      "args": []
    },
    {
      "file": "/drone/src/lib/private/legacy/OC_Template.php",
      "line": 267,
      "function": "printPage",
      "class": "OC\\Template\\Base",
      "type": "->",
      "args": []
    },
    {
      "file": "/drone/src/core/Controller/SetupController.php",
      "line": 106,
      "function": "printGuestPage",
      "class": "OC_Template",
      "type": "::",
      "args": [
        "",
        "installation",
        {
          "adminlogin": "",
          "adminpass": "",
          "dbuser": "",
          "dbpass": "",
          "dbname": "",
          "0": "And 13 more entries, set log level to debug to see all entries"
        }
      ]
    },
    {
      "file": "/drone/src/core/Controller/SetupController.php",
      "line": 85,
      "function": "display",
      "class": "OC\\Core\\Controller\\SetupController",
      "type": "->",
      "args": [
        "*** sensitive parameters replaced ***"
      ]
    },
    {
      "file": "/drone/src/lib/base.php",
      "line": 939,
      "function": "run",
      "class": "OC\\Core\\Controller\\SetupController",
      "type": "->",
      "args": [
        "*** sensitive parameters replaced ***"
      ]
    },
    {
      "file": "/drone/src/index.php",
      "line": 36,
      "function": "handleRequest",
      "class": "OC",
      "type": "::",
      "args": []
    }
  ],
  "File": "/drone/src/lib/private/Log/ErrorHandler.php",
  "Line": 92,
  "CustomMessage": "--"
}
```

juliushaertl replied:
I guess we do not load any scripts there

come-nc replied:
This will only happen on PHP < 7.4: https://www.php.net/manual/fr/function.array-merge.php

So not a problem for Nextcloud 24, but needs to be fixed if this will be backported to 23 or lower.

mejo- replied:
Thanks for the pointer, should be fixed now:

```diff
--- a/lib/public/Util.php
+++ b/lib/public/Util.php
@@ -237,7 +237,7 @@ class Util {
                uksort(self::$scripts, $sortByDeps);
 
                // Flatten array and remove duplicates
-               return array_unique(array_merge(...array_values(self::$scripts)));
+               return self::$scripts ? array_unique(array_merge(...array_values(self::$scripts))) : [];
        }
 
        /**
```
